app/src/main/java/ch/epfl/culturequest/notifications/FireMessaging.java

```diff
@@ -51,40 +53,45 @@ public static CompletableFuture<String> getDeviceToken() {
      * @return a CompletableFuture that will be completed with a boolean indicating if the
      * notification was sent successfully
      */
-    public static CompletableFuture<AtomicBoolean> sendNotification(String uid, PushNotification notification) {
-        // TODO: create instead an array of futures and return a future of array of booleans
-        CompletableFuture<AtomicBoolean> future = new CompletableFuture<>();
+    public static List<CompletableFuture<AtomicBoolean>> sendNotification(String uid, PushNotification notification) {
+        List<CompletableFuture<AtomicBoolean>> futures = new ArrayList<>();
 
         // add notification to database first
         Database.addNotification(uid, notification);
 
         // send the notification to all devices of the user
         Database.getDeviceTokens(uid).whenComplete((deviceTokens, throwable) -> {
             if (throwable != null || deviceTokens.isEmpty()) {
-                future.complete(new AtomicBoolean(false));
+                futures.add(CompletableFuture.completedFuture(new AtomicBoolean(false)));
                 return;
             }
 
-            for (String token : deviceTokens) {
-                JsonObject payload = buildNotificationPayload(token, notification);
+            // initialize the futures list with the correct number of futures
+            for (int i = 0; i < deviceTokens.size(); i++) {
+                futures.add(new CompletableFuture<>());
+            }
+
+            for (int i = 0; i < deviceTokens.size(); i++) {
+                JsonObject payload = buildNotificationPayload(deviceTokens.get(i), notification);
+                int finalI = i;
```

Comment:
Is this really necessary? Can't you just use `i` ?

Reply:
no due to future requirements it needs to use a final int that is not changing in a loop :/